prov/usnic/src/usdf_poll.c

```diff
@@ -0,0 +1,277 @@
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <fi.h>
+#include <fi_enosys.h>
+#include <fi_util.h>
+
+#include "usdf.h"
+#include "usdf_poll.h"
+
+static int usdf_poll_poll(struct fid_poll *pollset, void **context, int count)
+{
+	struct usdf_poll *_poll;
+	struct dlist_entry *item;
+	struct fid_list_entry *entry;
+	int ret = 0;
+	int copied = 0;
+	int pending;
+
+	USDF_TRACE_SYS(DOMAIN, "\n");
+
+	if (!pollset || !context) {
+		USDF_WARN_SYS(DOMAIN, "pollset and context can't be NULL.\n");
+		return -FI_EINVAL;
+	}
+
+	_poll = poll_ftou(pollset);
+
+	fastlock_acquire(&_poll->lock);
+
+	dlist_foreach(&_poll->list, item) {
+		entry = container_of(item, struct fid_list_entry, entry);
+		switch (entry->fid->fclass) {
+		case FI_CLASS_CQ:
+			pending = usdf_check_progress(entry->fid);
+			if ((pending == FI_SUCCESS) && (copied < count))
+				context[copied++] = entry->fid->context;
+			else if ((pending < 0) && (pending != -FI_EAGAIN))
+				ret = pending;
+			break;
+		default:
+			USDF_WARN_SYS(DOMAIN, "invalid object\n");
+			break;
+		}
+	}
+
+	fastlock_release(&_poll->lock);
+
+	if ((copied == count) || (copied > 0))
+		return copied;
+
+	return ret;
+}
+
+static int usdf_poll_add(struct fid_poll *pollset, struct fid *event_fid,
+		uint64_t flags)
+{
+	struct usdf_poll *_poll;
+	struct usdf_cq *_cq;
+	int ret;
+
+	USDF_TRACE_SYS(DOMAIN, "\n");
+
+	if (!pollset || !event_fid) {
+		USDF_WARN_SYS(DOMAIN, "pollset and event_fid can't be NULL.\n");
+		return -FI_EINVAL;
+	}
+
+	_poll = poll_ftou(pollset);
+
+	ret = fid_list_insert(&_poll->list, &_poll->lock, event_fid);
+	if (ret)
+		return ret;
+
+	switch (event_fid->fclass) {
```

Comment:
> I think this switch should move above the insert otherwise non-CQ fids end up on the list even after we return -FI_EINVAL.

Yeah it should be, I'll fix that.